Proxysql v2.3.2 crash

This is the place to report a **reproducible bug**, **documentation error** or **feature request** for ProxySQL.

Support questions will not be answered here. For community support, use the Google forum: https://groups.google.com/forum/#!forum/proxysql
* Note that due to time constraints ProxySQL engineers may be slow to respond to forum posts
* For information on paid support contracts: https://proxysql.com/services/support/

If you are submitting a reproducible bug report, please provide:
- [ ] A clear description of the issue
- [x] ProxySQL version       
          v2.3.2
- [x] OS version                
         FROM debian:stretch
- [ ] The steps to reproduce the issue
        I don't konw 
- [x] The **full** ProxySQL error log (default location: `/var/lib/proxysql/proxysql.log`)

```
{"log":"2025-08-08 05:50:46 main.cpp:1258:my_terminate(): [ERROR] ProxySQL crashed due to exception\n","stream":"stderr","time":"2025-08-07T21:50:46.395721907Z"}
{"log":"BACKTRACE ------------\n","stream":"stderr","time":"2025-08-07T21:50:46.398371178Z"}
{"log":"proxysql(_Z12my_terminatev+0x68) [0x5649655dd3a8]\n","stream":"stderr","time":"2025-08-07T21:50:46.398379908Z"}
{"log":"/usr/lib/x86_64-linux-gnu/libstdc++.so.6(+0x8f066) [0x7f7aa19c7066]\n","stream":"stderr","time":"2025-08-07T21:50:46.398382378Z"}
{"log":"/usr/lib/x86_64-linux-gnu/libstdc++.so.6(+0x8f0b1) [0x7f7aa19c70b1]\n","stream":"stderr","time":"2025-08-07T21:50:46.398384798Z"}
{"log":"/usr/lib/x86_64-linux-gnu/libstdc++.so.6(+0x8f2c9) [0x7f7aa19c72c9]\n","stream":"stderr","time":"2025-08-07T21:50:46.398387088Z"}
{"log":"/usr/lib/x86_64-linux-gnu/libstdc++.so.6(_ZSt19__throw_logic_errorPKc+0x3f) [0x7f7aa19ef8ef]\n","stream":"stderr","time":"2025-08-07T21:50:46.398388758Z"}
{"log":"proxysql(_ZN10httpserver9webserver13post_iteratorEPv13MHD_ValueKindPKcS4_S4_S4_S4_mm+0x76f) [0x564965ae257f]\n","stream":"stderr","time":"2025-08-07T21:50:46.398390718Z"}
{"log":"proxysql(+0x7a8b48) [0x564965b0fb48]\n","stream":"stderr","time":"2025-08-07T21:50:46.398392709Z"}
{"log":"proxysql(MHD_post_process+0x334) [0x564965b10384]\n","stream":"stderr","time":"2025-08-07T21:50:46.398394549Z"}
{"log":"proxysql(_ZN10httpserver9webserver27requests_answer_second_stepEP14MHD_ConnectionPKcS4_S4_PmPNS_7details14modded_requestE+0x8b) [0x564965ae481b]\n","stream":"stderr","time":"2025-08-07T21:50:46.398396279Z"}
{"log":"proxysql(_ZN10httpserver9webserver20answer_to_connectionEPvP14MHD_ConnectionPKcS5_S5_S5_PmPS1_+0x298) [0x564965ae4ae8]\n","stream":"stderr","time":"2025-08-07T21:50:46.398398699Z"}
{"log":"proxysql(+0x79d3ac) [0x564965b043ac]\n","stream":"stderr","time":"2025-08-07T21:50:46.398400689Z"}
{"log":"proxysql(+0x79e291) [0x564965b05291]\n","stream":"stderr","time":"2025-08-07T21:50:46.398402729Z"}
{"log":"proxysql(+0x7a2de8) [0x564965b09de8]\n","stream":"stderr","time":"2025-08-07T21:50:46.398409939Z"}
{"log":"proxysql(+0x7a3b4a) [0x564965b0ab4a]\n","stream":"stderr","time":"2025-08-07T21:50:46.398412069Z"}
{"log":"/lib/x86_64-linux-gnu/libpthread.so.0(+0x74a4) [0x7f7aa22e74a4]\n","stream":"stderr","time":"2025-08-07T21:50:46.398414059Z"}
{"log":"/lib/x86_64-linux-gnu/libc.so.6(clone+0x3f) [0x7f7aa1166d0f]\n","stream":"stderr","time":"2025-08-07T21:50:46.398416119Z"}
{"log":"----------------------\n","stream":"stderr","time":"2025-08-07T21:50:46.39844608Z"}
{"log":"Error: signal 6:\n","stream":"stderr","time":"2025-08-07T21:50:46.398470811Z"}
{"log":"proxysql(_Z13crash_handleri+0x2a)[0x5649655e625a]\n","stream":"stderr","time":"2025-08-07T21:50:46.398569864Z"}
{"log":"/lib/x86_64-linux-gnu/libc.so.6(+0x33060)[0x7f7aa10b1060]\n","stream":"stderr","time":"2025-08-07T21:50:46.398575085Z"}
{"log":"/lib/x86_64-linux-gnu/libc.so.6(gsignal+0xcf)[0x7f7aa10b0fff]\n","stream":"stderr","time":"2025-08-07T21:50:46.398595195Z"}
{"log":"/lib/x86_64-linux-gnu/libc.so.6(abort+0x16a)[0x7f7aa10b242a]\n","stream":"stderr","time":"2025-08-07T21:50:46.398597585Z"}
{"log":"/usr/lib/x86_64-linux-gnu/libstdc++.so.6(+0x8f06b)[0x7f7aa19c706b]\n","stream":"stderr","time":"2025-08-07T21:50:46.398611886Z"}
{"log":"/usr/lib/x86_64-linux-gnu/libstdc++.so.6(+0x8f0b1)[0x7f7aa19c70b1]\n","stream":"stderr","time":"2025-08-07T21:50:46.398664108Z"}
{"log":"/usr/lib/x86_64-linux-gnu/libstdc++.so.6(+0x8f2c9)[0x7f7aa19c72c9]\n","stream":"stderr","time":"2025-08-07T21:50:46.398667618Z"}
{"log":"/usr/lib/x86_64-linux-gnu/libstdc++.so.6(_ZSt19__throw_logic_errorPKc+0x3f)[0x7f7aa19ef8ef]\n","stream":"stderr","time":"2025-08-07T21:50:46.398710869Z"}
{"log":"proxysql(_ZN10httpserver9webserver13post_iteratorEPv13MHD_ValueKindPKcS4_S4_S4_S4_mm+0x76f)[0x564965ae257f]\n","stream":"stderr","time":"2025-08-07T21:50:46.398939577Z"}
{"log":"proxysql(+0x7a8b48)[0x564965b0fb48]\n","stream":"stderr","time":"2025-08-07T21:50:46.399157264Z"}
{"log":"proxysql(MHD_post_process+0x334)[0x564965b10384]\n","stream":"stderr","time":"2025-08-07T21:50:46.399379811Z"}
{"log":"proxysql(_ZN10httpserver9webserver27requests_answer_second_stepEP14MHD_ConnectionPKcS4_S4_PmPNS_7details14modded_requestE+0x8b)[0x564965ae481b]\n","stream":"stderr","time":"2025-08-07T21:50:46.399609339Z"}
{"log":"proxysql(_ZN10httpserver9webserver20answer_to_connectionEPvP14MHD_ConnectionPKcS5_S5_S5_PmPS1_+0x298)[0x564965ae4ae8]\n","stream":"stderr","time":"2025-08-07T21:50:46.399827137Z"}
{"log":"proxysql(+0x79d3ac)[0x564965b043ac]\n","stream":"stderr","time":"2025-08-07T21:50:46.400044325Z"}
{"log":"proxysql(+0x79e291)[0x564965b05291]\n","stream":"stderr","time":"2025-08-07T21:50:46.400263092Z"}
{"log":"proxysql(+0x7a2de8)[0x564965b09de8]\n","stream":"stderr","time":"2025-08-07T21:50:46.40049713Z"}
{"log":"proxysql(+0x7a3b4a)[0x564965b0ab4a]\n","stream":"stderr","time":"2025-08-07T21:50:46.400708728Z"}
{"log":"/lib/x86_64-linux-gnu/libpthread.so.0(+0x74a4)[0x7f7aa22e74a4]\n","stream":"stderr","time":"2025-08-07T21:50:46.400712248Z"}
{"log":"/lib/x86_64-linux-gnu/libc.so.6(clone+0x3f)[0x7f7aa1166d0f]\n","stream":"stderr","time":"2025-08-07T21:50:46.400729258Z"}
{"log":" ---- proxysql(_Z13crash_handleri+0x2a) [0x5649655e625a] : crash_handler(int)\n","stream":"stderr","time":"2025-08-07T21:50:46.403016636Z"}
{"log":" ---- /usr/lib/x86_64-linux-gnu/libstdc++.so.6(_ZSt19__throw_logic_errorPKc+0x3f) [0x7f7aa19ef8ef] : std::__throw_logic_error(char const*)\n","stream":"stderr","time":"2025-08-07T21:50:46.403022756Z"}
{"log":" ---- proxysql(_ZN10httpserver9webserver13post_iteratorEPv13MHD_ValueKindPKcS4_S4_S4_S4_mm+0x76f) [0x564965ae257f] : httpserver::webserver::post_iterator(void*, MHD_ValueKind, char const*, char const*, char const*, char const*, char const*, unsigned long, unsigned long)\n","stream":"stderr","time":"2025-08-07T21:50:46.403025326Z"}
{"log":" ---- proxysql(_ZN10httpserver9webserver20answer_to_connectionEPvP14MHD_ConnectionPKcS5_S5_S5_PmPS1_+0x298) [0x564965ae4ae8] : httpserver::webserver::answer_to_connection(void*, MHD_Connection*, char const*, char const*, char const*, char const*, unsigned long*, void**)\n","stream":"stderr","time":"2025-08-07T21:50:46.403028066Z"}
{"log":"Aborted (core dumped)\n","stream":"stderr","time":"2025-08-07T21:50:46.84603567Z"}
{"log":"2025-08-08 05:50:47 [INFO] Using config file /etc/proxysql.cnf\n","stream":"stderr","time":"2025-08-07T21:50:47.106388433Z"}
{"log":"2025-08-08 05:50:47 [INFO] Current RLIMIT_NOFILE: 1048576\n","stream":"stderr","time":"2025-08-07T21:50:47.106448985Z"}
{"log":"2025-08-08 05:50:47 [INFO] Using OpenSSL version: OpenSSL 1.1.1j  16 Feb 2021\n","stream":"stderr","time":"2025-08-07T21:50:47.106451585Z"}
{"log":"2025-08-08 05:50:47 [INFO] SSL keys/certificates found in datadir (/var/lib/proxysql): loading them.\n","stream":"stderr","time":"2025-08-07T21:50:47.106453585Z"}
{"log":"2025-08-08 05:50:47 [INFO] ProxySQL version 2.3.2-10-g8cd66cf\n","stream":"stderr","time":"2025-08-07T21:50:47.148219871Z"}
{"log":"2025-08-08 05:50:47 [INFO] Detected OS: Linux 10-53-12-245 4.19.0-6.el7.ucloud.x86_64 #1 SMP Wed Feb 12 07:32:16 UTC 2020 x86_64\n","stream":"stderr","time":"2025-08-07T21:50:47.148230991Z"}
{"log":"2025-08-08 05:50:47 [INFO] ProxySQL SHA1 checksum: 1d3d51e55d2ae2fd364c3c15b923fd02e5623180\n","stream":"stderr","time":"2025-08-07T21:50:47.148238431Z"}
{"log":"2025-08-08 05:50:47 [INFO] Loaded built-in SQLite3\n","stream":"stderr","time":"2025-08-07T21:50:47.148240541Z"}
{"log":"Standard ProxySQL MySQL Logger rev. 2.0.0714 -- MySQL_Logger.cpp -- Thu Sep 30 21:22:46 2021\n","stream":"stderr","time":"2025-08-07T21:50:47.150779318Z"}
{"log":"Standard ProxySQL Cluster rev. 0.4.0906 -- ProxySQL_Cluster.cpp -- Thu Sep 30 21:22:46 2021\n","stream":"stderr","time":"2025-08-07T21:50:47.151264974Z"}
{"log":"Standard ProxySQL Statistics rev. 1.4.1027 -- ProxySQL_Statistics.cpp -- Thu Sep 30 21:22:46 2021\n","stream":"stderr","time":"2025-08-07T21:50:47.151669498Z"}
{"log":"Standard ProxySQL HTTP Server Handler rev. 1.4.1031 -- ProxySQL_HTTP_Server.cpp -- Thu Sep 30 21:22:46 2021\n","stream":"stderr","time":"2025-08-07T21:50:47.152191366Z"}
{"log":"2025-08-08 05:50:47 ProxySQL_Admin.cpp:6178:flush_mysql_variables___database_to_runtime(): [WARNING] Impossible to set not existing variable mysql-query_processor_regex with value \"2\". Deleting. If the variable name is correct, this version doesn't support it\n","stream":"stderr","time":"2025-08-07T21:50:47.165198879Z"}
{"log":"Standard ProxySQL Admin rev. 2.0.6.0805 -- ProxySQL_Admin.cpp -- Thu Sep 30 21:22:46 2021\n","stream":"stderr","time":"2025-08-07T21:50:47.16638729Z"}
{"log":"2025-08-08 05:50:47 [INFO] ProxySQL SHA1 checksum: 1d3d51e55d2ae2fd364c3c15b923fd02e5623180\n","stream":"stderr","time":"2025-08-07T21:50:47.16639395Z"}
{"log":"Standard MySQL Threads Handler rev. 0.2.0902 -- MySQL_Thread.cpp -- Thu Sep 30 21:22:46 2021\n","stream":"stderr","time":"2025-08-07T21:50:47.16639599Z"}
{"log":"Standard MySQL Authentication rev. 0.2.0902 -- MySQL_Authentication.cpp -- Thu Sep 30 21:22:46 2021\n","stream":"stderr","time":"2025-08-07T21:50:47.16639776Z"}
```



If this is a crashing bug, please also include:
- [x] The package used to install ProxySQL
 wget https://github.com/sysown/proxysql/releases/download/v2.3.2/proxysql_2.3.2-debian9_amd64.deb
- [x] The compressed proxysql binary
- [x] The compressed core dump (Note: if you're worried it may contain sensitive data, please contact us for information on sharing it securely: https://proxysql.com/contact-us/)

[core.zip](https://github.com/user-attachments/files/21676433/core.zip)

If the above information is not provided, this issue is likely to be closed.

Please use markdown to format any code or SQL: https://guides.github.com/features/mastering-markdown/

Thank you!

--- a/lib/ProxySQL_RESTAPI_Server.cpp
+++ b/lib/ProxySQL_RESTAPI_Server.cpp
@@ -306,15 +306,32 @@ public:
 	}
 
 	const std::shared_ptr<http_response> render_POST(const http_request& req) {
-		std::string params=req.get_content();
+		std::string req_path = req.get_path();
+		std::string req_body = req.get_content();
+		std::string content_type = req.get_header(http::http_utils::http_header_content_type);
+
+		// validate Context-Type header
+		// reject unsupported content types
+		if (!content_type.empty() &&
+			content_type.find("application/json") == std::string::npos &&
+			content_type.find("text/json") == std::string::npos &&
+			content_type.find("text/plain") == std::string::npos) {
+#ifdef DEBUG
+			proxy_debug(PROXY_DEBUG_RESTAPI, 1, "Rejected POST - req: '%s', content-type: '%s', req_body: `%s`\n",
+			                                              req_path.c_str(), content_type.c_str(), req_body.c_str());
+#endif
+			auto response = std::shared_ptr<http_response>(new string_response(
+				"{\"error\": \"Unsupported Content-Type\"}",
+				http::http_utils::http_unsupported_media_type
+			));
+			add_headers(response);
+			return response;
+		}
 
 #ifdef DEBUG
-		const char* req_path { req.get_path().c_str() };
-		const char* p_params { params.c_str() };
-		proxy_debug(PROXY_DEBUG_RESTAPI, 1, "Processing POST - req: '%s', params: `%s`\n", req_path, p_params);
+		proxy_debug(PROXY_DEBUG_RESTAPI, 1, "Processing POST - req: '%s', params: `%s`\n", req_path.c_str(), req_body.c_str());
 #endif
-
-		return process_request(req, params);
+		return process_request(req, req_body);
 	}
 
 };
@@ -352,9 +369,15 @@ ProxySQL_RESTAPI_Server::ProxySQL_RESTAPI_Server(
 	// in the current mode concurrency on serving requests is low, and throughput is directly related with
 	// the time required to execute the target script, since each of the calls are blocking.
 #if defined(__FreeBSD__) || defined(__APPLE__)
-	ws = std::unique_ptr<httpserver::webserver>(new webserver(create_webserver(p).start_method(http::http_utils::start_method_T::THREAD_PER_CONNECTION)));
+	ws = std::unique_ptr<httpserver::webserver>(new webserver(
+		create_webserver(p)
+			.no_post_process()
+			.start_method(http::http_utils::start_method_T::THREAD_PER_CONNECTION)
+	));
 #else
-	ws = std::unique_ptr<httpserver::webserver>(new webserver(create_webserver(p)));
+	ws = std::unique_ptr<httpserver::webserver>(new webserver(
+		create_webserver(p).no_post_process()
+	));
 #endif
 	// NOTE: Enable for benchmarking purposes. In this mode each request will be served by it's own thread.
 	// ws = std::unique_ptr<httpserver::webserver>(new webserver(create_webserver(p).start_method(http::http_utils::start_method_T::THREAD_PER_CONNECTION)));
@@ -392,4 +415,3 @@ void ProxySQL_RESTAPI_Server::init() {
 void ProxySQL_RESTAPI_Server::print_version() {
     fprintf(stderr,"Standard ProxySQL REST API Server Handler rev. %s -- %s -- %s\n", PROXYSQL_RESTAPI_SERVER_VERSION, __FILE__, __TIMESTAMP__);
 }
-

